src/main/scala/com/fulcrumgenomics/bam/EstimatePoolingFractions.scala

```diff
@@ -0,0 +1,211 @@
+      logger.error("# insufficient coverage at some loci and/or too few SNPs being used.            #")
+      logger.error("# Please try again with a higher min-coverage-per-sample.                       #")
+      logger.error("#################################################################################")
+      fail(1)
+    }
+  }
+
+  /** Verify a provided sample list, or if none provided retrieve the set of samples from the VCF. */
+  private def pickSamplesToUse(vcfReader: VCFFileReader): Array[String] = {
+    if (samples.nonEmpty) {
+      val samplesInVcf = vcfReader.getFileHeader.getSampleNamesInOrder.toSet
+      val missingSamples = samples.filterNot(samplesInVcf.contains)
+      if (missingSamples.nonEmpty) fail(s"Samples not present in VCF: ${missingSamples.mkString(", ")}")
+      else samples.toArray.sorted
+    }
+    else {
+      vcfReader.getFileHeader.getSampleNamesInOrder.toSeq.toArray.sorted // toSeq.toArray is necessary cos util.ArrayList.toArray() exists
+    }
+  }
+
+  /** Loads up and merges all the interval lists provided. Returns None if no intervals were specified. */
+  private def loadIntervals: Option[IntervalList] = {
+    if (this.intervals.isEmpty) {
+      None
+    }
+    else {
+      val list = IntervalList.fromFile(this.intervals.head.toFile)
+      val dict = list.getHeader.getSequenceDictionary
+      this.intervals.tail.foreach { f =>
+        val tmp = IntervalList.fromFile(f.toFile)
+        if (!SequenceUtil.areSequenceDictionariesEqual(dict, tmp.getHeader.getSequenceDictionary))
+          fail(s"Sequence dictionaries differ between ${this.intervals.head} and ${f}")
+        else
+          list.addall(tmp.getIntervals)
+      }
+
+      Some(list.uniqued(false))
+    }
+  }
+
+  /** Generates an iterator over non-filtered bi-allelic SNPs where all the required samples are genotyped. */
+  def constructVcfIterator(in: VCFFileReader, intervals: Option[IntervalList], samples: Array[String]): Iterator[VariantContext] = {
+    val iterator = intervals match {
+      case None     => in.toIterator
+      case Some(is) => ByIntervalListVariantContextIterator(in, is)
+    }
+
+    val samplesAsUtilSet = CollectionUtil.makeSet(samples:_*)
+
+    iterator
+      .filterNot(_.isFiltered)
+      .map(_.subContextFromSamples(samplesAsUtilSet, true))
+      .filter(v => (v.isSNP && v.isBiallelic) || v.isMonomorphicInSamples)
+      .filter(_.getNoCallCount == 0)
+      .filter(v => v.getGenotypesOrderedByName.forall(gt => gt.getGQ >= this.minGenotypeQuality))
+  }
+
+  /** Constructs a SamLocusIterator that will visit every locus in the input. */
+  def constructBamIterator(loci: Traversable[Locus]): Iterator[LocusInfo] = {
+    val in = SamReaderFactory.make().open(this.bam)
+    val intervals = new IntervalList(in.getFileHeader)
+    loci.foreach(l => intervals.add(new Interval(l.chrom, l.pos, l.pos)))
+    val iterator = new SamLocusIterator(in, intervals.uniqued())
```

Comment:
I don't remember if this will include soft-clipped bases...I think not, but better be sure.